record-layer.go

```diff
@@ -116,47 +132,55 @@ func (r *RecordLayer) Rekey(epoch Epoch, factory aeadFactory, key []byte, iv []b
 		return err
 	}
 	r.cipher = cipher
+	if r.datagram && r.direction == directionRead {
+		r.readCiphers[epoch] = cipher
+	}
 	return nil
 }
 
-func (c *cipherState) formatSeq(datagram bool) []byte {
-	seq := append([]byte{}, c.seq...)
+func (r *RecordLayer) DiscardReadKey(epoch Epoch) {
+	if !r.datagram {
+		return
+	}
+
+	_, ok := r.readCiphers[epoch]
+	assert(ok)
+	delete(r.readCiphers, epoch)
+}
+
+func (c *cipherState) combineSeq(datagram bool) uint64 {
+	seq := c.seq
 	if datagram {
-		seq[0] = byte(c.epoch >> 8)
-		seq[1] = byte(c.epoch & 0xff)
+		seq |= uint64(c.epoch) << 48
 	}
 	return seq
 }
 
-func (c *cipherState) computeNonce(seq []byte) []byte {
+func (c *cipherState) computeNonce(seq uint64) []byte {
 	nonce := make([]byte, len(c.iv))
 	copy(nonce, c.iv)
 
-	offset := len(c.iv) - len(seq)
-	for i, b := range seq {
-		nonce[i+offset] ^= b
+	s := seq
+
+	offset := len(c.iv)
+	for i := 0; i < 8; i++ {
+		nonce[(offset-i)-1] ^= byte(s & 0xff)
+		s >>= 8
 	}
+	logf(logTypeCrypto, "Computing nonce for sequence # %x -> %x", seq, nonce)
 
 	return nonce
 }
 
 func (c *cipherState) incrementSequenceNumber() {
-	var i int
-	for i = len(c.seq) - 1; i >= 0; i-- {
-		c.seq[i]++
-		if c.seq[i] != 0 {
-			break
-		}
-	}
-
-	if i < 0 {
+	if c.seq >= 1<<48 {
```

Comment:
If you've got a 48-bit field, you should panic before you overflow, i.e., `c.seq >= (1 << 48) - 1`.